Fix mypy invocation in noxfile

leandro-lucarella-frequenz opened the pull request:
When invoking `mypy`, the `src` directory (`frequenz.sdk` package) isn't being checked because 2 bugs:

* The `src` path is excluded
* A transformation of `src` into a package name was a ill done copy&paste from the channels repository, and said `frequenz.channels` instead of `frequenz.sdk`

This commit fixes this issue by reworking how the conversion between paths and packages needed by `mypy` is done. Now the conversion is more generic and explicit, providing a list of paths and a mapping between paths and packages.

This is a bit more verbose to declare but it can be reused more safely and gives some extra flexibility, which can be handy when we move this to a general tooling repo.

We also use package names for plain python files (like `noxfile.py`), so we don't need to invoke `mypy` twice.

llucax committed:
nox: Use __future__ annotations

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
nox: Fix bug when invoking mypy

When invoking mypy, the src directory (frequenz.sdk package) isn't
being checked because 2 bugs:

* The "src" path is excluded
* A transformation of src into a package name was a ill done copy&paste
  from the channels repository, and said frequenz.channels instead of
  frequenz.sdk

This commit fixes this issue by reworking how the conversion between
paths and packages needed by mypy is done. Now the conversion is more
generic and explicit, providing a list of paths and a mapping between
paths and packages.

This is a bit more verbose to declare but it can be reused more safely
and gives some extra flexibility, which can be handy when we move this
to a general tooling repo.

We also use package names for plain python files (like noxfile.py), so
we don't need to invoke mypy twice.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
This is a draft because now `mypy` fails with some errors that were never reported before, I'll fix them at some point if nobody else did (feel free to jump in and push more commits fixing `mypy` issues).

shsms commented:
I'd say we add exceptions for the affected files in the noxfile and get this in,  so that new PRs get checked.  And fix the broken files in subsequent PRs.

leandro-lucarella-frequenz commented:
Mmm, not sure about that, I think we at least get some type checking on the affected files indirectly from other packages (like `tests`), if we add exceptions then we don't get any type checking for the affected files anymore. Not sure what is worse. I would try to fix the current issues ASAP and be done with it.

leandro-lucarella-frequenz commented:
If it's too difficult or takes too long, I agree we can go with the exception option.

llucax committed:
Add some missing type hints

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Remove use of Any

We need to use an explicit elif checking the type too to make sure the
type checker gets the proper types for np.concatenate().

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Use SupportsIndex for __{get,set}item__

This is the correct type for this.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Workaround some mypy issues

There seem to be 2 different mypy bugs at play while forwarding the call
to __setitem__().

First we need to check that the combination of input arguments are
correct to make the type checker happy (I guess it could be inferred
from the @overloads, but it's not currently working without this hack).

Then we need to ignore a no-untyped-call error, for some reason it
can't get the type for self._buffer.__setitem__().

We need to use `SupportsFloat` to handle cases where an `int` is used
instead of a float, that's used in the unit tests. Apparently `int` the
type checker accepts an `int` where a `float` is required but checking
`ininstance(value, float)` fails if `value` is an `int`.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Fix type checking for MovingWindow

* Add overloads for __getitem__() and use SupportsIndex.
* Because of the above, we need to register the class as a subclass of
  Sequence like this because otherwise type-checking fails complaining
  that MovingWindow has more overloads of __getitem__() than Sequence
  (which doesn't have an overload with a datetime key)
* Add missing generic parameter to Task and Sequence.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Add a workaround for a mypy issue

See https://github.com/python/mypy/issues/14774 for more details.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Add missing generic parameter

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Convert timeouts to int

The microgrid API takes the timeouts as int instead of float.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Deal with missing type info in microgrid API calls

grpc.aio doesn't have type information, and there are all sorts of
tricks in it, so some funtions that look like they are not async really
are (because they return a awaitable). For example [`unary_unary()`]
returns a [`UnaryUnaryMultiCallable`] which returns a [`UnaryUnaryCall`]
which is `awaitable`.

Also there seem to be wrong type info regarding `SetBounds()`, saying
`request_iterator` is a required argument but it is not.
[`stream_unary()`] returns a [`StreamUnaryMultiCallable`] which takes an
optional `request_iterator`.

[`unary_unary()`]: https://github.com/grpc/grpc/blob/v1.51.3/src/python/grpcio/grpc/aio/_base_channel.py#L271-L276
[`UnaryUnaryMultiCallable`]: https://github.com/grpc/grpc/blob/v1.51.3/src/python/grpcio/grpc/aio/_base_channel.py#L28-L41
[`UnaryUnaryCall`]: https://github.com/grpc/grpc/blob/v1.51.3/src/python/grpcio/grpc/aio/_base_call.py#L138-L149
['stream_unary()']: https://github.com/grpc/grpc/blob/v1.51.3/src/python/grpcio/grpc/aio/_base_channel.py#L313-L318
[`StreamUnaryMultiCallable`]: https://github.com/grpc/grpc/blob/v1.51.3/src/python/grpcio/grpc/aio/_base_channel.py#L106-L118

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Remove NOP "# type: ignore"

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Fix wrong type hint

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Don't reuse variable to store different types

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Ignore type checking for networkx

There is a stubs package available, but it's not working:
https://github.com/eggplants/networkx-stubs/issues/1

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Import class from where it is defined

The ResamplerConfig is defined in timeseries._resampling and not
explicitly exported by actor._resampling.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
I pushed all fixes that don't require a deep knowledge on the affected modules to fix. The only missing fixes are in:

* `src/frequenz/sdk/microgrid/client/_client.py`: @ela-kotulska-frequenz maybe you can have a look at this one?

  * There are only 2 missing errors that I don't know how to fix and look like proper bugs, as they are calls to functions without passing all required arguments.
  * The changes I made also needs a review because there are some fishy things, like most calls being done with `await` to functions that are not `async`.

* `src/frequenz/sdk/timeseries/_formula_engine/*`: This is the package (several modules are affected) with the most errors and which look the harder to fix, because there is a lot of generic code with complicated types, this probably needs to wait until @sahas-subramanian-frequenz is back from holidays as it probably will take too much time for anybody else to fix it. We should probably add these files to some `mypy` exclusion to be able to merge this PR before this is fixed.

llucax committed:
Exclude _formula_engine from type checking

This is to be able to merge the easy fixes until we have time to fix
this package, as it is quite complicated.

For this to work, we also need to install the package as editable for
`the ci_checks_max` nox session, otherwise the exclusion rule doesn't
work as it is not found in `src/`.

We also take the opportunity to make all installs in nox editable, it is
usually what we want while developing.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
I just pushed a commit to exclude the formula engine from `mypy` and created an issue to fix it separately in the future.

* https://github.com/frequenz-floss/frequenz-sdk-python/issues/226

Since this affects the whole code base, I would like to merge it ASAP, hope you can have a look at the microgrid API client next week @ela-kotulska-frequenz

mathias-baumann-frequenz commented:
Needs rebase

leandro-lucarella-frequenz commented:
> Needs rebase

Done! I increased the priority to urgent and removed the draft status because if we keep merging PRs we might be as well introducing more (uncaught) typing errors.

Ideally we should merge this before any other PR, although it still misses the fix for the 2 remaining microgrid API errors.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_ringbuffer.py:
```diff
@@ -220,14 +220,17 @@ def window(
-            return window
+                    raise TypeError(
+                        "Bad type, something seems to have went wrong with type checking"
+                    )
```
I think this change will need to update the docstring to add `TypeError` in the `Raises` session

leandro-lucarella-frequenz replied:
I think this is actually a good reason to make it an `assert`, I didn't want to use `assert False`, but it doesn't make sense to document an error that will only happen if type checking is not performed.

daniel-zullo-frequenz replied:
Right, I see your point. Though I don't know if the nox-docstrings checks will complain about it.

leandro-lucarella-frequenz replied:
No, it doesn't.

daniel-zullo-frequenz commented on noxfile.py:
```diff
@@ -50,13 +50,29 @@
+"""A list of path to be used by default and its corresponding package name.
+
+The package name is needed for mypy, as it takes packages when full import
+checking needs to be doneg
```
typo in `doneg`

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_ringbuffer.py:
```diff
@@ -422,10 +425,22 @@ def __setitem__(
+        elif isinstance(index_or_slice, slice) and isinstance(value, Iterable):
+            self._buffer.__setitem__(index_or_slice, value)  # type: ignore[no-untyped-call]
+        else:
+            raise TypeError("Incompatible input arguments")
```
`TypeError` might need to be added in the docstring `Raises` session

leandro-lucarella-frequenz replied:
Same as above.

daniel-zullo-frequenz reviewed (commented):
I have a couple of minor comments to check for, LGTM otherwise

ela-kotulska-frequenz commented on noxfile.py:
```diff
@@ -182,23 +191,13 @@ def mypy(session: nox.Session, install_deps: bool = True) -> None:
-        pkg_args.append("-p")
-        pkg_args.append(pkg)
+    packages = (DEFAULT_PATH_PACKAGES[i] for i in _source_file_paths(session))
+    pkg_args = _flatten(("-p", p) for p in packages)
```
Maybe just:
pkg_args = (("-p", p) for p in DEFAULT_PATH_PACKAGES.values())

leandro-lucarella-frequenz replied:
I don't get this. We need to *flatten* the `("-p", p)` tuple to then pass it as a list of arguments to `mypy`. But I think you found a bug, not sure if this is what you meant. I'm translating the output of `_source_file_paths(session)` but this could give just the positional arguments to `nox -s mypy -- args`, so it will fail in that case :unamused:

leandro-lucarella-frequenz replied:
I fixed this bug but I'm still unsure if this is what you saw that was wrong or it was something else :)

ela-kotulska-frequenz replied:
It looks good now :) Just that you check  session.posargs: in two places (in line 189, for the first time)
```
    # If we have session arguments, we just use those...
    if session.posargs:
        session.run("mypy", *common_args, *session.posargs)
        return'
```

leandro-lucarella-frequenz replied:
🤦‍♂️

That explains why it actually worked when passing positional args :shrug:

ela-kotulska-frequenz replied:
Hahahaha xD 
You seems to be so tired of this PR, that you started to belief in things that just happen in magic way :D

leandro-lucarella-frequenz replied:
Fixed.

leandro-lucarella-frequenz replied:
> You seems to be so tired of this PR, that you started to belief in things that just happen in magic way :D

YES!

![Get off of me!](https://media.tenor.com/NRHJPgOH9EYAAAAC/pilot-hardy-ew-get-off-of-me.gif)

llucax committed:
Add back tomli dependency for pydocstyle

This is needed so pydocstyle can read configuration from pyproject.toml.
This is mainly as a precaution as there is currently no pydocstyle in
the pyproject.toml file, but it would be very unexpected if some is
added and then is not picked up by pydocstyle.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
Updated fixing daniz comments and the last microgrid api issues. I also had to change a few commits because `mypy` didn't complain but `pylint` did, or test were failing.

Diff here: https://github.com/frequenz-floss/frequenz-sdk-python/compare/633c36f6e5639dd96534050ad5f89196b373e9f7..0c5316660cff4955722348698c21b883a7a585a1

leandro-lucarella-frequenz commented:
Mmm, tests are still failing, it seems like the ignore for `formula_engine` is not working, but it is working locally. I need to look into it :unamused:

leandro-lucarella-frequenz commented:
OK, I think I found the issue and it should work now. It was a problem with the `ci_checks_max` `nox` session, that is only used in the CI. This didn't install the package as editable, so the exclusion rule didn't work as it uses paths, and `src` was not found if the package is installed normally.

leandro-lucarella-frequenz commented:
OK, new issue comes from adding pyarrow-stubs, it seems it can't find a matching dependency for python 3.10 and 3.11 :unamused:

leandro-lucarella-frequenz commented:
OK, it is finally passing. I had to remove `pyarrow-stubs` because it wasn't supported in Python 3.11 (and the repo didn't existed anymore, it looked a bit fishy).

ela-kotulska-frequenz commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -212,7 +212,7 @@ async def components(self) -> Iterable[Component]:
             component_list = await self.api.ListComponents(
                 microgrid_pb.ComponentFilter(),
-                timeout=DEFAULT_GRPC_CALL_TIMEOUT,
+                timeout=int(DEFAULT_GRPC_CALL_TIMEOUT),
```
How did you find that timeout is `int`?

I am looking and the definition and it is float:
https://github.com/grpc/grpc/blob/v1.51.3/src/python/grpcio/grpc/aio/_base_channel.py#L31

leandro-lucarella-frequenz replied:
I don't remember exactly, but it was very likely because I was hinted by `mypy`. Maybe it also gets the type info wrong as with the `request_iterator`... :thinking: 

I will try to revert it and see what's complaining about and add a `type: ignore` if it's wrong.

Thanks for checking!

leandro-lucarella-frequenz replied:
```
src/frequenz/sdk/microgrid/client/_client.py:227: error: Argument "timeout" to "__call__" of "UnaryUnaryMultiCallable" has incompatible type "float"; expected "Optional[int]"  [arg-type]
```

I'll add an ignore :+1:

ela-kotulska-frequenz commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -266,8 +278,13 @@ async def connections(
-                    connection_filter, timeout=int(DEFAULT_GRPC_CALL_TIMEOUT)
+                # grpc.aio is missing types and mypy thinks this is not
+                # awaitable, but it is
+                cast(
+                    Awaitable[Any],
+                    self.api.ListConnections(
+                        connection_filter, timeout=int(DEFAULT_GRPC_CALL_TIMEOUT)
+                    ),
```
Did you try:
`Awaitable[Any] -> Awaitable[microgrid_pb.ConnectionList]` ?

leandro-lucarella-frequenz replied:
No, I was a bit lazy to try to figure out what was the correct type (and looking at types for pb is torture :scream:), I just wanted to let `gather` know that it was an awaitable.

I will try, thanks!

leandro-lucarella-frequenz replied:
Works! :+1:

ela-kotulska-frequenz commented on pyproject.toml:
```diff
@@ -136,6 +136,12 @@ src_paths = ["src", "examples", "tests"]
 asyncio_mode = "auto"
 required_plugins = [ "pytest-asyncio", "pytest-mock" ]
 
+[tool.mypy]
+# This is a temporary hack, for details see:
+# * https://github.com/frequenz-floss/frequenz-sdk-python/pull/219
+# * https://github.com/frequenz-floss/frequenz-sdk-python/issues/226
+exclude = ["src/frequenz/sdk/timeseries/_formula_engine/"]
```
From type checking of from tests?

leandro-lucarella-frequenz replied:
I don't get this. This is excluding the `_formula_engine` subdirectory because it has a lot of errors that need to be fixed in a separate PR. This is not related to tests.

ela-kotulska-frequenz replied:
Ok, then it is fine.
I thought you wanted to quiet the tests because they are failing randomly.
But it is definitely better to keep them and quiet mypy in the _formula_engine

llucax committed:
Remove pyarrow dependency

The modules that depended on it were recently removed.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
Pushed the suggested fixes and also added a commit to remove the `pyarrow` dependency, as it was only used in the modules recently removed.
